lib/kino/utils/table.ex

```diff
@@ -1,77 +1,75 @@
+  # Common functions for handling various Elixir terms
+  # as table records.
+
+  @type record :: map() | list({term(), term()}) | tuple() | term()
 
   @doc """
-  Computes table column keys that accommodate for all the given records.
+  Computes table column specifications that accommodate
+  the given records.
+
+  Note that the columns are computed based on the first
+  record, if present.
   """
-  def keys_for_records(records) do
+  @spec columns_for_records(list(record())) :: list(Kino.Table.column())
+  def columns_for_records(records) do
     case Enum.at(records, 0) do
-      nil ->
-        []
-
-      first_record ->
-        first_record_keys = keys_for_record(first_record)
-
-        all_keys =
-          records
-          |> Enum.reduce(MapSet.new(), fn record, keys ->
-            record
-            |> keys_for_record()
-            |> MapSet.new()
-            |> MapSet.union(keys)
-          end)
-          |> MapSet.to_list()
-          |> Enum.sort()
-
-        # If all records have the same structure, keep the order,
-        # otherwise return the sorted accumulated keys
-        if length(first_record_keys) == length(all_keys) do
-          first_record_keys
-        else
-          all_keys
-        end
+      nil -> []
+      first_record -> columns_for_record(first_record)
     end
   end
 
-  defp keys_for_record(record) when is_tuple(record) do
+  defp columns_for_record(record) when is_tuple(record) do
     record
     |> Tuple.to_list()
     |> Enum.with_index()
     |> Enum.map(&elem(&1, 1))
+    |> keys_to_columns()
   end
 
-  defp keys_for_record(record) when is_map(record) do
+  defp columns_for_record(record) when is_map(record) do
     if schema = ecto_schema(record) do
-      schema.__schema__(:fields)
+      columns_for_schema(schema)
     else
-      record |> Map.keys() |> Enum.sort()
+      record |> Map.keys() |> Enum.sort() |> keys_to_columns()
     end
   end
 
-  defp keys_for_record(record) when is_list(record) do
-    record
-    |> Keyword.keys()
+  defp columns_for_record(record) when is_list(record) do
+    record |> Keyword.keys() |> keys_to_columns()
   end
 
-  defp keys_for_record(_record) do
+  defp columns_for_record(_record) do
     # If the record is neither of the expected enumerables,
     # we treat it as a single column value
-    [:item]
+    keys_to_columns([:item])
   end
 
   @doc """
   Converts keys to column specifications.
   """
+  @spec keys_to_columns(list(term())) :: list(Kino.Table.column())
   def keys_to_columns(keys) do
     Enum.map(keys, fn key -> %{key: key, label: inspect(key)} end)
   end
 
+  @doc """
+  Computes table column specifications for the given Ecto schema.
+  """
+  @spec columns_for_schema(module()) :: list(Kino.Table.column())
+  def columns_for_schema(schema) do
+    for field <- schema.__schema__(:fields) do
+      type = schema.__schema__(:type, field)
+      %{key: field, label: inspect(field), type: to_string(type)}
```

Comment:
Composite and parameterized types can be tuples. :)

Reply:
Good point, so in that case `inspect` should look fine, right?

![image](https://user-images.githubusercontent.com/17034772/153202090-029e9077-f2dd-4e4d-b08f-16ed33af7e92.png)